leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java

```diff
@@ -137,22 +138,23 @@ public T buildResponse(final Response coapResponse) {
         endpoint.sendRequest(coapRequest);
     }
 
-    public synchronized void cancelPendingRequests(String registrationId) {
-        ArrayList<Request> requests = pendingRequests.get(registrationId);
+    public void cancelPendingRequests(String registrationId) {
+        List<Request> requests = pendingRequests.remove(registrationId);
         if (requests != null) {
             for (Request coapRequest : requests) {
                 coapRequest.cancel();
-                removePendingRequest(registrationId, coapRequest);
             }
         }
     }
 
-    private synchronized void addPendingRequest(String registrationId, Request coapRequest) {
+    private void addPendingRequest(String registrationId, Request coapRequest) {
         // Get or create all pending requests for this registrationId
-        ArrayList<Request> requests = pendingRequests.get(registrationId);
+        List<Request> requests = pendingRequests.get(registrationId);
         if (requests == null) {
-            requests = new ArrayList<>();
-            pendingRequests.put(registrationId, requests);
+            requests = Collections.synchronizedList(new ArrayList<Request>());
```

Comment:
This is not thread safe.  2 threads could check that requests == null and so add the new collections twice. (Only one request will be added.